chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2193 @@
+        block = _read_block(radio, i, 0x20)
+        data += block
+        _do_status(radio, i)
+    _do_status(radio, radio._memsize)
+    LOG.info("done.")
+
+    if append_model:
+        data += radio.MODEL.ljust(8)
+
+    # LOG.debug("done.")
+    return memmap.MemoryMapBytes(data)
+
+
+def _exit_write_block(radio):
+    serial = radio.pipe
+    try:
+        serial.write(b"E")
+
+    except:
+        raise errors.RadioError("Radio refused to exit programming mode")
+
+
+def _write_block(radio, addr, data):
+    serial = radio.pipe
+    cmd = struct.pack(">cHb", b'W', addr, 0x20)
+    data = radio.get_mmap()[addr + 8: addr + 40]
+    # The checksum needs to be in the last
+    check_sum = bytes([sum(data) & 0xFF])
+    data += check_sum
+    used_data = cmd + data
+    serial.write(used_data)
+
+    ack = radio.pipe.read(1)
+    if ack != b"\x06":
+        raise errors.RadioError("Radio refused to accept block 0x%04x" % addr)
+
+
+def _do_upload(radio):
+    ident = _ident_radio(radio)
+    radio_version = _get_radio_firmware_version(radio)
+    LOG.info("Radio Version is %s" % repr(radio_version))
+
+    # Main block
+    LOG.debug("Uploading...")
+
+    for start_addr, end_addr in radio._ranges_main:
+        for addr in range(start_addr, end_addr, 0x20):
+            _write_block(radio, addr, 0x20)
+            _do_status(radio, addr)
+    _exit_write_block(radio)
+    LOG.debug("Upload all done.")
+
+
+TX_POWER = [chirp_common.PowerLevel("Low",  watts=1.00),
+            chirp_common.PowerLevel("Mid",  watts=4.00),
+            chirp_common.PowerLevel("High", watts=8.00)]
+
+TDH8_CHARSET = chirp_common.CHARSET_UPPER_NUMERIC + \
+    "!@#$%^&*()+-=[]:\";'<>?,./"
+
+
+@directory.register
+class TDH8(chirp_common.CloneModeRadio):
+    """TIDRADIO TD-H8"""
+    VENDOR = "TIDRADIO"
+    MODEL = "TD-H8"
+    ident_mode = b'P31183\xff\xff'
+    BAUD_RATE = 38400
+    _memsize = 0x1eef
+    _ranges_main = [(0x0000, 0x1eef)]
+    # _other_main = [(0x1868, 0x1eff)]
+    _idents = [TD_H8]
+    _vhf_range = (136000000, 175000000)
+    # _220_range = (220000000, 260000000)
+    _uhf_range = (400000000, 521000000)
+    _aux_block = True
+    _tri_power = True
+    _gmrs = False
+    _ham = False
+    _mem_params = (0x1F2F)  # poweron_msg offset
+
+    # offset of fw version in image file
+    _fw_ver_file_start = 0x1838
+    _fw_ver_file_stop = 0x1846
+    NEEDS_COMPAT_SERIAL = False
+    _valid_chars = TDH8_CHARSET
+
+    @classmethod
+    def get_prompts(cls):
+        rp = chirp_common.RadioPrompts()
+        rp.experimental = \
+            ('Due to the fact that the manufacturer continues to '
+             'release new versions of the firmware with obscure and '
+             'hard-to-track changes, this driver may not work with '
+             'your device. Thus far and to the best knowledge of the '
+             'author, no TD-H8 radios have been harmed by using CHIRP. '
+             'However, proceed at your own risk!')
```

Comment:
Do you still want this to be displayed to the user? Aren't you associated with the manufacturer? :)